src/seedu/addressbook/data/person/Address.java

```diff
@@ -10,23 +14,37 @@
 
     public static final String EXAMPLE = "123, some street";
     public static final String MESSAGE_ADDRESS_CONSTRAINTS = "Person addresses can be in any format";
-    public static final String ADDRESS_VALIDATION_REGEX = ".+";
+    public static final String ADDRESS_VALIDATION_REGEX = "(\\w+)(, ([^,]+)?(, ([^,]+)?(, (\\d+))?)?)?";
 
     public final String value;
     private boolean isPrivate;
-
+    private Block block;
+    private Unit unit;
+    private PostalCode postalCode;
+    private Street street;
+    
     /**
      * Validates given address.
      *
      * @throws IllegalValueException if given address string is invalid.
      */
     public Address(String address, boolean isPrivate) throws IllegalValueException {
+        
         String trimmedAddress = address.trim();
-        this.isPrivate = isPrivate;
-        if (!isValidAddress(trimmedAddress)) {
+        Pattern pattern = Pattern.compile(ADDRESS_VALIDATION_REGEX);
+        Matcher matcher = pattern.matcher(trimmedAddress);
+        
+        if(matcher.find()){
+            this.block = new Block(matcher.group(1));
+            this.street = new Street(matcher.group(3));
+            this.unit = new Unit(matcher.group(5));
+            this.postalCode = new PostalCode(matcher.group(7));
+            this.isPrivate = isPrivate;
+        }
+        else {
             throw new IllegalValueException(MESSAGE_ADDRESS_CONSTRAINTS);
         }
-        this.value = trimmedAddress;
+        this.value = this.toString();
```

Comment:
No need to save the input address in `this.value` as it may cause inconsistency due to duplication.